src/main/java/org/opentripplanner/model/impl/OtpTransitServiceImpl.java

```diff
@@ -144,174 +111,90 @@
-    }
-
-    @Override
-    public List<StopTime> getStopTimesForTrip(Trip trip) {
-
-        if (stopTimesByTrip == null) {
-            stopTimesByTrip = getAllStopTimes().stream().collect(groupingBy(StopTime::getTrip));
-
-            for (List<StopTime> stopTimes : stopTimesByTrip.values()) {
-                Collections.sort(stopTimes);
-            }
+    private Map<FeedScopedId, List<ShapePoint>> mapShapePoints(Collection<ShapePoint> shapePoints) {
+        Map<FeedScopedId, List<ShapePoint>> map = shapePoints.stream()
+                .collect(groupingBy(ShapePoint::getShapeId));
+        for (List<ShapePoint> list : map.values()) {
+            Collections.sort(list);
         }
-
-        return nullSafeUnmodifiableList(stopTimesByTrip.get(trip));
+        return map;
     }
 
-    @Override
-    public List<FeedScopedId> getAllServiceIds() {
-        ensureCalendarDatesByServiceIdRelation();
-        ensureCalendarsByServiceIdRelation();
-        Set<FeedScopedId> serviceIds = new HashSet<>();
-        serviceIds.addAll(calendarDatesByServiceId.keySet());
-        serviceIds.addAll(calendarsByServiceId.keySet());
-        return new ArrayList<>(serviceIds);
-    }
-
-    @Override
-    public List<ServiceCalendarDate> getCalendarDatesForServiceId(FeedScopedId serviceId) {
-        ensureCalendarDatesByServiceIdRelation();
-        return nullSafeUnmodifiableList(calendarDatesByServiceId.get(serviceId));
-    }
-
-    @Override
-    public ServiceCalendar getCalendarForServiceId(FeedScopedId serviceId) {
-        ensureCalendarsByServiceIdRelation();
-        List<ServiceCalendar> calendars = calendarsByServiceId.get(serviceId);
-
-        if(calendars == null || calendars.isEmpty()) {
-            return null;
-        }
-        if(calendars.size() == 1) {
-            return calendars.get(0);
-        }
-        throw new MultipleCalendarsForServiceIdException(serviceId);
-    }
-
-
-    /*  Private Methods */
-
-    private void ensureCalendarDatesByServiceIdRelation() {
-        if (calendarDatesByServiceId == null) {
-            calendarDatesByServiceId = getAllCalendarDates().stream()
-                    .collect(groupingBy(ServiceCalendarDate::getServiceId));
-        }
-    }
-
-    private void ensureCalendarsByServiceIdRelation() {
-        if (calendarsByServiceId == null) {
-            calendarsByServiceId = getAllCalendars().stream()
-                    .collect(groupingBy(ServiceCalendar::getServiceId));
-        }
-    }
-
-    private void ensureShapePointRelation() {
-        if (shapePointsByShapeId == null) {
-            shapePointsByShapeId = getAllShapePoints().stream()
-                    .collect(groupingBy(ShapePoint::getShapeId));
-            for (List<ShapePoint> list : shapePointsByShapeId.values()) {
-                Collections.sort(list);
-            }
+    private static <T> List<T> nullSafeUnmodifiableList(Collection<T> c) {
```

Comment:
`list = new ArrayList<>(c)` creates a new List with all elements passed in.

Reply:
It actually convert a Set to a List.

Reply:
Oh right, I somehow missed that `c` parameter. So it will handle any kind of collection. But then it is not null-safe in the sense that it will tolerate `c == null` right? This ArrayList constructor's Javadoc says: c – the collection whose elements are to be placed into this list. Throws: NullPointerException – if the specified collection is null

Reply:
You are right, the null-safety is probably survived trough some refactorings - I will rename the method: `immutableList` and add some JavaDoc.